### PR TITLE
reduces chances of generating duplicate tokens in db

### PR DESCRIPTION
**Pull request by jmtcsngr**

### CODE REVIEW

**Comment by jmtcsngr** [comment]
Remove TODOs

File: lib/model.js [hunk 1]
```diff
@@ -107,6 +91,27 @@ function findMatchingTokens(token) {
...
+      return err;
+    }).then((count) => {
+      return count === 0 ? Promise.resolve(token) :
+                           Promise.reject('Duplicated token');
+    }, err => {
+      return Promise.reject(err);
+    });
+    return uniqueTokenPromise;
+  }, err => {
+    return err;
```
Comment (andrew-nowak) [inline lib/model.js]: Should be `return Promise.reject(err);`

File: lib/model.js [hunk 2]
```diff
@@ -107,6 +91,27 @@ function findMatchingTokens(token) {
   });
 }
 
+function generateTokenPromise() {
+  // TODO refactor to prevent collisions with tokens in db
+  let tokenStringPromise = tokenUtils.generateTokenStringPromise().then((token) => {
+    let uniqueTokenPromise = p_collection.then((collection) => {
+      return collection.find({token: token}).count();
+    }, err => {
+      return err;
```
Comment (andrew-nowak) [inline lib/model.js]: Should be `return Promise.reject(err);`

File: test/server/model.spec.js [hunk 3]
```diff
@@ -159,6 +159,52 @@ describe('model', function() {
...
+
+        let user = 'user@example.com';
+        let p_insert = model.createToken(user, 'test creation');
+        let p_insert2 = p_insert.then(() => {
+          expect(utils.generateTokenStringPromise.calls.count()).toEqual(1);
+          let secondTokenPromise = model.createToken(user, 'test creation');
+          expect(utils.generateTokenStringPromise.calls.count()).toEqual(2);
+          return secondTokenPromise;
+        }, (err) => {
+          done.fail(err);
```
Comment (andrew-nowak) [inline test/server/model.spec.js]: Can be replaced with `}, done.fail);`

File: test/server/model.spec.js [hunk 4]
```diff
@@ -159,6 +159,52 @@ describe('model', function() {
...
+                reject(err);
+              }
+            });
+          });
+
+          p_count.then(function(count) {
+            expect(count).toBe(1);
+            done();
+          }, (err) => {
+            done.fail(err);
```
Comment (andrew-nowak) [inline test/server/model.spec.js]: Can be replaced with `}, done.fail);`

File: CHANGES [hunk 5]
```diff
@@ -1,9 +1,11 @@
+<<<<<<< HEAD
```
Comment (andrew-nowak) [inline CHANGES]: merge conflict remnant :(

File: lib/model.js [hunk 6]
```diff
@@ -154,14 +169,13 @@ function generateTokenPromise() {
...
  */
-function findMatchingTokens(token) {
-  return p_collection.then(function(collection) {
-    logger.debug('Searching database for token: ' + token);
-    let query = {token: token};
-    // This will automatically reject if collection.find() throws an error.
-    // No need to manually catch and reject.
-    return collection.find(query);
-  });
+function getDocument(p_cursor) {
```
Comment (andrew-nowak) [inline lib/model.js]: getDocument has the jsdoc for findMatchingTokens

File: lib/model.js [hunk 7]
```diff
@@ -174,23 +188,6 @@ function findMatchingTokens(token) {
...
-
-/**
- * Checks that a cursor only contains one result. Returned promise rejects if
- * there was no results or more than one.
- *
- * @param {Promise} p_cursor - takes value of a mongodb.Cursor when fulfilled.
- * @return {Promise} Fulfills if only one token exists in the db, otherwise
- *   rejects with a {@link DbError}.
- */
 function cursorHasExactlyOneDoc(p_cursor) {
```
Comment (andrew-nowak) [inline lib/model.js]: cursorHasExactlyOneDoc has the jsdoc for getDocument, and cursorHasExactlyOneDoc's jsdoc seems to have been deleted

File: lib/model.js [hunk 8]
```diff
@@ -217,6 +214,7 @@ function _buildToken(token, user, justification) {
   let validDuration = moment.duration(
     constants.TOKEN_DURATION, constants.TOKEN_DURATION_UNIT
   );
+  // TODO move to constants
```
Comment (andrew-nowak) [inline lib/model.js]: is this still true?

File: lib/model.js [hunk 9]
```diff
@@ -127,25 +127,40 @@ p_collection.catch(function(reason) {
...
-    // 24 bytes will generate a 32 character base64 string
-    logger.debug('Generating new token');
-    randomBytes(24, function(err, buf) {
-      if (err) {
-        reject(err);
-      }
-      // http://stackoverflow.com/questions/8838624/node-js-send-email-on-
-      // registration/8842959#8842959
-      resolve(buf.toString('base64').replace(/\//g, '_').replace(/\+/g, '-'));
+  // TODO refactor to prevent collisions with tokens in db
```
Comment (andrew-nowak) [inline lib/model.js]: Is this still TODO?

**Review by andrew-nowak — COMMENTED** [review]
Something odd happened with the jsdocs here...

**Pull request merged**